File: modules/java/main/src/test/java/com/mooltiverse/oss/nyx/services/git/util/FileSystemUtil.java
/*
 * Copyright 2020 Mooltiverse
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mooltiverse.oss.nyx.services.git.util;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.OutputStream;
import java.nio.file.Files;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

/**
 * File system utilities used in tests
 */
public class FileSystemUtil {
    /**
     * Default constructor is private on purpose.
     */
    private FileSystemUtil() {
        super();
    }

    /**
     * Returns a string with the hierarchical representation of the directory contents
     * 
     * @param directory the directory to render
     * 
     * @return a string representation of the given directory
     */
    public static String renderDirectoryTree(File directory) {
        return renderDirectoryTree(directory, Integer.MAX_VALUE);
    }

    /**
     * Prints the hierarchical representation of the directory contents to the given output stream.
     * The output stream is flushed once written.
     * 
     * @param directory the directory to render
     * @param out the stream to write to
     */
    public static void renderDirectoryTree(File directory, OutputStream out)
        throws Exception {
        renderDirectoryTree(directory, Integer.MAX_VALUE, out);
    }

    /**
     * Prints the hierarchical representation of the directory contents to the given output stream.
     * The output stream is flushed once written.
     * 
     * @param directory the directory to render
     * @param maxDepth the maximum depth of objects to render (starting from 1 with the given directory)
     * @param out the stream to write to
     */
    public static void renderDirectoryTree(File directory, int maxDepth, OutputStream out)
        throws Exception {
        out.write(renderDirectoryTree(directory, maxDepth).getBytes());
        out.flush();
    }

    /**
     * Returns a string with the hierarchical representation of the directory contents.
     * 
     * @param directory the directory to render
     * @param maxDepth the maximum depth of objects to render (starting from 1 with the given directory)
     * 
     * @return a string representation of the given directory
     */
    public static String renderDirectoryTree(File directory, int maxDepth) {
        StringBuilder sb = new StringBuilder();
        sb.append(directory.getAbsolutePath());
        sb.append(System.getProperty("line.separator"));
        if (directory.isDirectory()) {
            File[] items = directory.listFiles();
            // first render directories
            for (int i=0; i<items.length; i++)
                if (items[i].isDirectory()) renderDirectoryItem(sb, "", items[i], maxDepth-1, i==items.length-1);
            // then render files
            for (int i=0; i<items.length; i++)
                if (items[i].isFile()) renderDirectoryItem(sb, "", items[i], maxDepth-1, i==items.length-1);
        }

        return sb.toString();
    }

    /**
     * Returns a string with the hierarchical representation of the directory item (file or directory). This method
     * is recursive and allows to render any level of depth.
     * 
     * @param sb the string builder to append the output to
     * @param lineStart is the initial part of the line
     * @param item the item to render
     * @param maxDepth the maximum depth of objects to render
     * @param last a boolean telling if this is the last item within the parent
     * 
     * @return a string representation of the given item
     */
    private static void renderDirectoryItem(StringBuilder sb, String lineStart, File item, int maxDepth, boolean last) {
        if (maxDepth > 0) {
            lineStart = lineStart + (last ? "   " : "   ");
            sb.append(lineStart);
            sb.append(item.getName());
            if (item.isFile()) {
                sb.append(" (");
                sb.append(item.length());
                sb.append(" bytes)");
            }
            sb.append(System.getProperty("line.separator"));
            if (item.isDirectory()) {
                File[] items = item.listFiles();
                for (int i=0; i<items.length; i++) {
                    renderDirectoryItem(sb, lineStart, items[i], maxDepth-1, i==items.length-1);
                }
            }
        }
    }

    /**
     * Returns all files in the given directory (and its sub directories).
     * 
     * @param directory the directory to list the files from.
     * @param subdirs if {@code true} also include subdirs.
     * 
     * @return the collection of files resulting from the search
     */
    public static Collection<File> getFiles(File directory, boolean subdirs) {
        return getFiles(directory, null, subdirs);
    }

    /**
     * Returns all files in the given directory (and its sub directories).
     * 
     * @param directory the directory to list the files from.
     * @param filters a collection of strings representing filters for file names not to be included.
     * If a file name is equal or {@link String#matches(String) matches} (using a regex) one of these strings
     * it will be skipped. If the collection is {@code null} or empty no filters will be applied.
     * @param subdirs if {@code true} also include subdirs.
     * 
     * @return the collection of files resulting from the search
     */
    public static Collection<File> getFiles(File directory, Collection<String> filters, boolean subdirs) {
        Collection<File> res = new ArrayList<File>();
        for (File f: directory.listFiles()) {
            if (Objects.isNull(filters)) {
                if (subdirs && f.isDirectory())
                    res.addAll(getFiles(f, filters, subdirs));
                else if (f.isFile())
                    res.add(f);
            } else {
                boolean include = true;
                for (String filter: filters) {
                    if (filter.equals(f.getName()) || f.getName().matches(filter)) {
                        include = false;
                        break;
                    }
                }
                if (include) {
                    if (subdirs && f.isDirectory())
                        res.addAll(getFiles(f, filters, subdirs));
                    else if (f.isFile())
                        res.add(f);
                }
            }
        }

        return res;
    }

    /**
     * Creates a new temporary file with a random name in the given directory.
     * 
     * @param dir the directory to create the file in.
     * @param suffix the file name prefix. If {@code null} the file will have a random prefix.
     * @param suffix the file extension. If {@code null} the file will have the {@code .tmp} extension.
     * 
     * @return the file that has been created
     * 
     * @throws Exception in case of any exception
     */
    public static File newTempFile(File dir, String prefix, String extension)
        throws Exception {
        return File.createTempFile(Objects.isNull(prefix) ? RandomUtil.randomAlphabeticString(3) : prefix, extension, dir);
    }

    /**
     * Creates a new temporary directory with a random name in the given directory.
     * 
     * @param dir the parent directory to create the subdirectory in. If {@code null} the default temporary directory is used
     * @param prefix an optional prefix for the directory name. It may be {@code null}
     * 
     * @return the directory that has been created
     * 
     * @throws Exception in case of any exception
     */
    public static File newTempDirectory(File parent, String prefix)
        throws Exception {
        File res = Objects.isNull(parent) ? Files.createTempDirectory(prefix).toFile() : Files.createTempDirectory(parent.toPath(), prefix).toFile();
        res.deleteOnExit();
        return res;
    }

    /**
     * Creates a new directory with a random name in the given directory.
     * 
     * @param dir the directory to create the subdirectory in.
     * 
     * @return the directory that has been created
     * 
     * @throws Exception in case of any exception
     */
    public static File newDirectory(File dir)
        throws Exception {
        File res = new File(dir.getAbsolutePath()+File.pathSeparator+RandomUtil.randomAlphabeticString(5));
        res.mkdirs();
        return res;
    }

    /**
     * Creates a number of new temporary files with both binary and text content in the given directory.
     * Binary files have the {@code .bin} extension and an increasing size depending on how
     * many are created.
     * Text files have the {@code .txt} extension and an increasing size depending on how
     * many are created.
     * 
     * @param dir the directory to create the file in.
     * @param count the number of files to create
     * 
     * @return the set of files that has been created
     * 
     * @throws Exception in case of any exception
     */
    public static Set<File> newTempFilesWithRandomContent(File dir, int count)
        throws Exception {
        Set<File> res = new HashSet<File>(count);
        for (int i=0; i< count; i++) {
            // alternate binary and text files
            res.add(i % 2 == 0 ? newTempFileWithRandomBytes(dir, "file", ".bin", 50*(i+1)) : newTempFileWithRandomText(dir, "file", ".txt", 5*(i+1), 10*(i+1)));
        }
        return res;
    }

    /**
     * Creates a number of new temporary files with text content in the given directory.
     * 
     * @param dir the directory to create the file in.
     * @param count the number of files to create
     * 
     * @return the set of files that has been created
     * 
     * @throws Exception in case of any exception
     */
    public static Set<File> newTempFilesWithRandomText(File dir, int count)
        throws Exception {
        Set<File> res = new HashSet<File>(count);
        for (int i=0; i< count; i++) {
            res.add(newTempFileWithRandomText(dir, "file", ".txt", 5*(i+1), 10*(i+1)));
        }
        return res;
    }

    /**
     * Creates a number of new temporary files with binary content in the given directory.
     * 
     * @param dir the directory to create the file in.
     * @param count the number of files to create
     * 
     * @return the set of files that has been created
     * 
     * @throws Exception in case of any exception
     */
    public static Set<File> newTempFilesWithRandomBytes(File dir, int count)
        throws Exception {
        Set<File> res = new HashSet<File>(count);
        for (int i=0; i< count; i++) {
            res.add(newTempFileWithRandomBytes(dir, "file", ".bin", 50*(i+1)));
        }
        return res;
    }

    /**
     * Creates a new temporaty file with a random name and binary content in the given directory.
     * 
     * @param dir the directory to create the file in.
     * @param suffix the file name prefix. If {@code null} the file will have a random prefix.
     * @param suffix the file extension. If {@code null} the file will have the {@code .tmp} extension.
     * @param length the number of bytes to write to the target file.
     * 
     * @return the file that has been created
     * 
     * @throws Exception in case of any exception
     */
    public static File newTempFileWithRandomBytes(File dir, String prefix, String extension, int length)
        throws Exception {
        File res = newTempFile(dir, prefix, extension);
        writeRandomBytes(res, length);
        return res;
    }

    /**
     * Creates a new temporary file with a random name and text content in the given directory.
     * 
     * @param dir the directory to create the file in.
     * @param suffix the file name prefix. If {@code null} the file will have a random prefix.
     * @param suffix the file extension. If {@code null} the file will have the {@code .tmp} extension.
     * @param lines the number of lines to write to the target file.
     * @param length the length of lines to write to the target file.
     * 
     * @return the file that has been created
     * 
     * @throws Exception in case of any exception
     */
    public static File newTempFileWithRandomText(File dir, String prefix, String extension, int lines, int length)
        throws Exception {
        File res = newTempFile(dir, prefix, extension);
        writeRandomText(res, lines, length);
        return res;
    }

    /**
     * Writes random binary content to the given files.
     * 
     * @param files the files to write to. If they doen't exist yet then they are created.
     * @param length the number of bytes to write to the target files.
     * 
     * @throws Exception in case of any exception
     */
    public static void writeRandomBytes(Collection<File> files, int length)
        throws Exception {
        for (File f: files)
            writeRandomBytes(f, length);
    }

    /**
     * Writes random text content to the given files.
     * 
     * @param files the files to write to. If they doen't exist yet then they are created.
     * @param lines the number of lines to write to the target files.
     * @param length the length of lines to write to the target files.
     * 
     * @throws Exception in case of any exception
     */
    public static void writeRandomText(Collection<File> files, int lines, int length)
        throws Exception {
        for (File f: files)
            writeRandomText(f, lines, length);
    }

    /**
     * Writes random binary content to the given file.
     * 
     * @param file the file to write to. If it doesn't exist yet then it is created.
     * @param length the number of bytes to write to the target file.
     * 
     * @throws Exception in case of any exception
     */
    public static void writeRandomBytes(File file, int length)
        throws Exception {
        if (!file.exists())
            file.createNewFile();
        
        FileOutputStream writer = new FileOutputStream(file);
        writer.write(RandomUtil.randomBytes(length));
        writer.flush();
        writer.close();
    }

    /**
     * Writes random text content to the given file.
     * 
     * @param file the file to write to. If it doesn't exist yet then it is created.
     * @param lines the number of lines to write to the target file.
     * @param length the length of lines to write to the target file.
     * 
     * @throws Exception in case of any exception
     */
    public static void writeRandomText(File file, int lines, int length)
        throws Exception {
        String[] contentLines = new String[lines];
        for (int i=0; i<lines; i++)
            contentLines[i] = RandomUtil.randomAlphabeticString(length);
        appendLines(file, contentLines);
    }

    /**
     * Appends the given lines to the end of the given file.
     * 
     * @param file the file to write to. If it doesn't exist yet then it is created.
     * @param lines the text lines to append 
     * 
     * @throws Exception in case of any exception
     */
    public static void appendLines(File file, String... lines)
        throws Exception {
        if (!file.exists())
            file.createNewFile();

        FileWriter fw = new FileWriter(file, true);
        BufferedWriter bw = new BufferedWriter(fw);
        for (String line: lines) {
            bw.write(line);
            bw.newLine();
        }
        bw.flush();
        bw.close();
    }
}

File: modules/go/nyx/test/integration/git/util/filesystem_util.go
/*
 * Copyright 2020 Mooltiverse
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package util

import (
	"io"            // https://pkg.go.dev/io
	"io/fs"         // https://pkg.go.dev/fs
	"math"          // https://pkg.go.dev/math
	"os"            // https://pkg.go.dev/os
	"path/filepath" // https://pkg.go.dev/path/filepath
	"strconv"       // https://pkg.go.dev/strconv
	"strings"       // https://pkg.go.dev/strings

	regexp2 "github.com/dlclark/regexp2" // https://pkg.go.dev/github.com/dlclark/regexp2, we need to use this instead of the standard 'regexp' to have support for lookarounds (look ahead), even if this implementation is a little slower
)

/*
Returns a string with the hierarchical representation of the directory contents

Arguments are as follows:

- directory the directory to render
*/
func RenderDirectoryTree(directory string) string {
	return RenderDirectoryTreeWithDepth(directory, math.MaxInt)
}

/*
Prints the hierarchical representation of the directory contents to the given output stream.

Arguments are as follows:

- directory the directory to render
- out the stream to write to
*/
func RenderDirectoryTreeTo(directory string, out io.Writer) {
	RenderDirectoryTreeWithDepthTo(directory, math.MaxInt, out)
}

/*
Prints the hierarchical representation of the directory contents to the given output stream.

Arguments are as follows:

- directory the directory to render
- maxDepth the maximum depth of objects to render (starting from 1 with the given directory)
- out the stream to write to
*/
func RenderDirectoryTreeWithDepthTo(directory string, maxDepth int, out io.Writer) {
	out.Write([]byte(RenderDirectoryTreeWithDepth(directory, maxDepth)))
}

/*
Returns a string with the hierarchical representation of the directory contents.

Arguments are as follows:

  - directory the directory to render
    maxDepth the maximum depth of objects to render (starting from 1 with the given directory)
*/
func RenderDirectoryTreeWithDepth(directory string, maxDepth int) string {
	var sb strings.Builder
	absolutePath, err := filepath.Abs(directory)
	if err != nil {
		panic(err)
	}
	sb.WriteString(absolutePath)
	sb.WriteString("\n")

	file, err := os.Open(absolutePath)
	if err != nil {
		panic(err)
	}
	fileInfo, err := file.Stat()
	if err != nil {
		panic(err)
	}
	if fileInfo.IsDir() {
		children, err := os.ReadDir(absolutePath)
		if err != nil {
			panic(err)
		}
		// first render directories
		for i, child := range children {
			if child.IsDir() {
				renderDirectoryItem(&sb, "", child, filepath.Join(absolutePath, child.Name()), maxDepth-1, i == len(children)-1)
			}
		}
		// then render files
		for i, child := range children {
			if !child.IsDir() {
				renderDirectoryItem(&sb, "", child, filepath.Join(absolutePath, child.Name()), maxDepth-1, i == len(children)-1)
			}
		}
	}

	return sb.String()
}

/*
Returns a string with the hierarchical representation of the directory item (file or directory). This method
is recursive and allows to render any level of depth.

Arguments are as follows:

- sb the string builder to append the output to
- lineStart is the initial part of the line
- item the item to render
- itemPath the full path of the item to render
- maxDepth the maximum depth of objects to render
- last a boolean telling if this is the last item within the parent
*/
func renderDirectoryItem(sb *strings.Builder, lineStart string, item fs.DirEntry, itemPath string, maxDepth int, last bool) {
	if maxDepth > 0 {
		lineStart := lineStart + "   "

		sb.WriteString(lineStart)
		sb.WriteString(item.Name())

		if !item.IsDir() {
			sb.WriteString(" (")
			itemInfo, err := item.Info()
			if err != nil {
				panic(err)
			}
			sb.WriteString(strconv.FormatInt(itemInfo.Size(), 10))
			sb.WriteString(" bytes)")
		}
		sb.WriteString("\n")
		if item.IsDir() {
			children, err := os.ReadDir(itemPath)
			if err != nil {
				panic(err)
			}
			for i, child := range children {
				renderDirectoryItem(sb, lineStart, child, filepath.Join(itemPath, child.Name()), maxDepth-1, i == len(children)-1)
			}
		}
	}
}

/*
Returns all files in the given directory (and its sub directories).

Arguments are as follows:

- directory the directory to list the files from.
- subdirs if true also include subdirs.
*/
func GetFiles(directory string, subdirs bool) []string {
	return GetFilesFiltered(directory, nil, subdirs)
}

/*
Returns all files in the given directory (and its sub directories).

Arguments are as follows:

  - directory the directory to list the files from.
  - filters a collection of strings representing filters for file names not to be included.
    If a file name is equal or matches (using a regex) one of these strings
    it will be skipped. If the collection is nil or empty no filters will be applied.
  - subdirs if true also include subdirs.
*/
func GetFilesFiltered(directory string, filters []string, subdirs bool) []string {
	res := []string{}
	children, err := os.ReadDir(directory)
	if err != nil {
		panic(err)
	}
	for _, f := range children {
		if filters == nil || len(filters) == 0 {
			if subdirs && f.IsDir() {
				res = append(res, GetFilesFiltered(filepath.Join(directory, f.Name()), filters, subdirs)...)
			} else {
				res = append(res, f.Name())
			}
		} else {
			include := true
			for _, filter := range filters {

				if filter == f.Name() {
					include = false
					break
				} else {
					re := regexp2.MustCompile(filter, 0)
					if isMatch, _ := re.MatchString(f.Name()); isMatch {
						include = false
						break
					}
				}
			}
			if include {
				if subdirs && f.IsDir() {
					res = append(res, GetFilesFiltered(filepath.Join(directory, f.Name()), filters, subdirs)...)
				} else {
					res = append(res, f.Name())
				}
			}
		}
	}

	return res
}

/*
Creates a new temporary file with a random name in the given directory.

Arguments are as follows:

- dir the directory to create the file in.
- prefix the file name prefix. If nil the file will have a random prefix.
*/
func NewTempFile(dir string, prefix *string) string {
	var prf string
	if prefix == nil {
		prf = RandomAlphabeticString(3, 11)
	} else {
		prf = *prefix
	}

	f, err := os.CreateTemp(dir, prf)
	if err != nil {
		panic(err)
	}
	defer os.Remove(f.Name())
	return f.Name()
}

/*
Creates a new temporary directory with a random name in the given directory.

Arguments are as follows:

- dir the parent directory to create the subdirectory in. If empty the default temporary directory is used
- prefix an optional prefix for the directory name. It may be nil
*/
func NewTempDirectory(parent string, prefix *string) string {
	prf := ""
	if prefix != nil {
		prf = *prefix
	}
	dir, err := os.MkdirTemp(parent, prf)
	if err != nil {
		panic(err)
	}

	return dir
}

/*
Creates a new directory with a random name in the given directory.

Arguments are as follows:

- dir the directory to create the subdirectory in.
*/
func NewDirectory(dir string) string {
	dirName := filepath.Join(dir, RandomAlphabeticString(5, 13))
	err := os.MkdirAll(dirName, 0750)
	if err != nil {
		panic(err)
	}
	return dirName
}

/*
Creates a number of new temporary files with both binary and text content in the given directory.

Arguments are as follows:

- dir the directory to create the file in.
- count the number of files to create
*/
func NewTempFilesWithRandomContent(dir string, count int) []string {
	res := make([]string, count)
	prefix := "file"
	for i := 0; i < count; i++ {
		// alternate binary and text files
		if i%2 == 0 {
			res[i] = NewTempFileWithRandomBytes(dir, &prefix, 50*(i+1))
		} else {
			res[i] = NewTempFileWithRandomText(dir, &prefix, 5*(i+1), 10*(i+1))
		}
	}
	return res
}

/*
Creates a number of new temporary files with text content in the given directory.

Arguments are as follows:

- dir the directory to create the file in.
- count the number of files to create
*/
func NewTempFilesWithRandomText(dir string, count int) []string {
	res := make([]string, count)
	prefix := "file"
	for i := 0; i < count; i++ {
		res[i] = NewTempFileWithRandomText(dir, &prefix, 5*(i+1), 10*(i+1))
	}
	return res
}

/*
Creates a number of new temporary files with binary content in the given directory.

Arguments are as follows:

- dir the directory to create the file in.
- count the number of files to create
*/
func NewTempFilesWithRandomBytes(dir string, count int) []string {
	res := make([]string, count)
	prefix := "file"
	for i := 0; i < count; i++ {
		res[i] = NewTempFileWithRandomBytes(dir, &prefix, 50*(i+1))
	}
	return res
}

/*
Creates a new temporaty file with a random name and binary content in the given directory.

Arguments are as follows:

- dir the directory to create the file in.
- suffix the file name prefix. If nil the file will have a random prefix.
- length the number of bytes to write to the target file.
*/
func NewTempFileWithRandomBytes(dir string, prefix *string, length int) string {
	res := NewTempFile(dir, prefix)
	WriteRandomBytesToFile(res, length)
	return res
}

/*
Creates a new temporary file with a random name and text content in the given directory.

Arguments are as follows:

- dir the directory to create the file in.
- suffix the file name prefix. If nil the file will have a random prefix.
- lines the number of lines to write to the target file.
- length the length of lines to write to the target file.
*/
func NewTempFileWithRandomText(dir string, prefix *string, lines int, length int) string {
	res := NewTempFile(dir, prefix)
	WriteRandomTextToFile(res, lines, length)
	return res
}

/*
Writes random binary content to the given files.

Arguments are as follows:

- files the files to write to. If they doen't exist yet then they are created.
- length the number of bytes to write to the target files.
*/
func WriteRandomBytesToFiles(files []string, length int) {
	for _, f := range files {
		WriteRandomBytesToFile(f, length)
	}
}

/*
Writes random text content to the given files.

Arguments are as follows:

- files the files to write to. If they doen't exist yet then they are created.
- lines the number of lines to write to the target files.
- length the length of lines to write to the target files.
*/
func WriteRandomTextToFiles(files []string, lines int, length int) {
	for _, f := range files {
		WriteRandomTextToFile(f, lines, length)
	}
}

/*
Writes random binary content to the given file.

Arguments are as follows:

- file the file to write to. If it doesn't exist yet then it is created.
- length the number of bytes to write to the target file.
*/
func WriteRandomBytesToFile(file string, length int) {
	f, err := os.OpenFile(file, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		panic(err)
	}
	_, err = f.Write(RandomBytes(length, 21))
	if err != nil {
		panic(err)
	}

	if err := f.Close(); err != nil {
		panic(err)
	}
}

/*
Writes random text content to the given file.

Arguments are as follows:

- file the file to write to. If it doesn't exist yet then it is created.
- lines the number of lines to write to the target file.
- length the length of lines to write to the target file.
*/
func WriteRandomTextToFile(file string, lines int, length int) {
	contentLines := make([]string, lines)
	for i := 0; i < lines; i++ {
		contentLines[i] = RandomAlphabeticString(length, 17)
	}
	AppendLines(file, contentLines...)
}

/*
Appends the given lines to the end of the given file.

Arguments are as follows:

- file the file to write to. If it doesn't exist yet then it is created.
- lines the text lines to append
*/
func AppendLines(file string, lines ...string) {
	f, err := os.OpenFile(file, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		panic(err)
	}
	for _, line := range lines {
		_, err := f.Write([]byte(line + "\n"))
		if err != nil {
			panic(err)
		}
	}
	if err := f.Close(); err != nil {
		panic(err)
	}
}
